libraries/AP_HAL_ChibiOS/Scheduler.cpp

```diff
@@ -493,14 +540,35 @@ void Scheduler::expect_delay_ms(uint32_t ms)
         return;
     }
     if (ms == 0) {
-        expect_delay_start = 0;
+        if (expect_delay_nesting > 0) {
+            expect_delay_nesting--;
+        }
+        if (expect_delay_nesting == 0) {
+            expect_delay_start = 0;
+        }
     } else {
-        expect_delay_start = AP_HAL::millis();
+        uint32_t now = AP_HAL::millis();
+        if (expect_delay_start != 0) {
+            // we already have a delay running, possibly extend it
```

Comment:
Shouldn't having to extend the delay be an internal error?  Any time we're crossing this code it means the "outside" (or "already running") delay wasn't conservative enough.